src/Worker/Core/Shims/TaskEntityShim.cs

```diff
@@ -0,0 +1,240 @@
+using Microsoft.DurableTask.Entities;
+using Microsoft.Extensions.Logging;
+using DTCore = DurableTask.Core;
+
+namespace Microsoft.DurableTask.Worker.Shims;
+
+/// <summary>
+/// Shim that provides the entity context and implements batched execution.
+/// </summary>
+class TaskEntityShim : DTCore.Entities.TaskEntity
+{
+    readonly DataConverter dataConverter;
+    readonly ITaskEntity taskEntity;
+    readonly EntityInstanceId entityId;
+
+    readonly StateShim state;
+    readonly ContextShim context;
+    readonly OperationShim operation;
+
+    // entities roll back the state and actions when the user code throws an unhandled exception
+    string? checkpointedState;
+    int checkpointedPosition;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="TaskEntityShim"/> class.
+    /// </summary>
+    /// <param name="dataConverter">The data converter.</param>
+    /// <param name="taskEntity">The task entity.</param>
+    /// <param name="entityId">The entity ID.</param>
+    public TaskEntityShim(
+        DataConverter dataConverter,
+        ITaskEntity taskEntity,
+        EntityId entityId)
+    {
+        this.dataConverter = Check.NotNull(dataConverter);
+        this.taskEntity = Check.NotNull(taskEntity);
+        this.entityId = new EntityInstanceId(entityId.Name, entityId.Key);
+        this.state = new StateShim(this, dataConverter);
+        this.context = new ContextShim(this, dataConverter);
+        this.operation = new OperationShim(this);
+    }
+
+    /// <inheritdoc />
+    public override async Task<EntityBatchResult> ExecuteOperationBatchAsync(EntityBatchRequest operations)
+    {
+        // initialize/reset the state and action list
+        this.state.CurrentState = operations.EntityState;
+        this.context.Rollback(0);
+
+        List<OperationResult> results = new List<OperationResult>();
+
+        for (int i = 0; i < operations.Operations!.Count; i++)
+        {
+            OperationRequest currentOperation = operations.Operations![i];
+            this.operation.SetNameAndInput(currentOperation.Operation!, currentOperation.Input);
+            this.Checkpoint();
+
+            try
+            {
+                object? result = await this.taskEntity.RunAsync(this.operation);
+                string? serializedResult = this.dataConverter.Serialize(result);
+                results.Add(new OperationResult() { Result = serializedResult });
+            }
+            catch (Exception applicationException)
+            {
+                results.Add(new OperationResult()
+                {
+                    Result = null,
+                    ErrorMessage = "exception in application code",
+                    FailureDetails = new FailureDetails(applicationException),
+                });
+
+                this.Rollback();
+            }
+        }
+
+        this.ClearCheckpoint(); // want to ensure the old state can be GC'd even if this shim is being cached
```

Comment:
If we want to manage GC, maybe all the shim objects can be created at the start of `ExecuteOperationBatchAsync` and then let them go out of scope and be GC'd entirely.

I personally don't think optimizing to re-use these objects between batches is worth the added complexity.

Reply:
I think it is worth it if the batches are large and the operations are very short.

Reply:
The nice part here is that the user does not really see these optimizations in any way, from the outside it looks clean.

Reply:
If you want to keep it the way it is, that is fine. Not my style, but I'll live 😄